Retry communication with Artifactory

I'm seen multiple occurrences of the following or similar recently. It would be nice if it retried more after waiting for a bit, rather than having to restart the build.
https://ci.eclipse.org/openj9/job/Build_JDK11_x86-64_linux_Personal/198
```
[Pipeline] getArtifactoryServer
[Pipeline] newBuildInfo
[Pipeline] artifactoryUpload
expected to call org.jfrog.hudson.pipeline.common.types.ArtifactoryServer.upload but wound up catching artifactoryUpload; see: https://jenkins.io/redirect/pipeline-cps-method-mismatches/
12:20:08  [consumer_0] Deploying artifact: https://140-211-168-230-openstack.osuosl.org/artifactory/ci-eclipse-openj9/Build_JDK11_x86-64_linux_Personal/198/OpenJ9-JDK11-x86-64_linux-20190821-154159.tar.gz
12:20:13  [consumer_1] Deploying artifact: https://140-211-168-230-openstack.osuosl.org/artifactory/ci-eclipse-openj9/Build_JDK11_x86-64_linux_Personal/198/native-test-libs.tar.gz
12:34:32  Error occurred for request PUT /artifactory/ci-eclipse-openj9/Build_JDK11_x86-64_linux_Personal/198/OpenJ9-JDK11-x86-64_linux-20190821-154159.tar.gz;build.parentNumber=617;build.parentName=Pipeline_Build_Test_JDK11_x86-64_linux;build.buildIdentifier=eclipse/openj9%236792;build.timestamp=1566402061113;build.name=Build_JDK11_x86-64_linux_Personal;build.number=198 HTTP/1.1: Read timed out.
12:34:32  Attempting retry #1
12:40:17  [consumer_0] An exception occurred during execution:
12:40:17  java.lang.RuntimeException: java.io.IOException: Failed to deploy file. Status code: 500 Response message: Artifactory returned the following errors: 
12:40:17   Status code: 500
12:40:17  	at org.jfrog.build.extractor.clientConfiguration.util.spec.SpecDeploymentConsumer.consumerRun(SpecDeploymentConsumer.java:44)
12:40:17  	at org.jfrog.build.extractor.producerConsumer.ConsumerRunnableBase.run(ConsumerRunnableBase.java:11)
12:40:17  	at java.lang.Thread.run(Thread.java:748)
12:40:17  Caused by: java.io.IOException: Failed to deploy file. Status code: 500 Response message: Artifactory returned the following errors: 
12:40:17   Status code: 500
12:40:17  	at org.jfrog.build.extractor.clientConfiguration.client.ArtifactoryBuildInfoClient.uploadFile(ArtifactoryBuildInfoClient.java:685)
12:40:17  	at org.jfrog.build.extractor.clientConfiguration.client.ArtifactoryBuildInfoClient.doDeployArtifact(ArtifactoryBuildInfoClient.java:367)
12:40:17  	at org.jfrog.build.extractor.clientConfiguration.client.ArtifactoryBuildInfoClient.deployArtifact(ArtifactoryBuildInfoClient.java:355)
12:40:17  	at org.jfrog.build.extractor.clientConfiguration.util.spec.SpecDeploymentConsumer.consumerRun(SpecDeploymentConsumer.java:39)
12:40:17  	... 2 more
12:40:17  
[Pipeline] }
[Pipeline] // dir
[Pipeline] }
[Pipeline] // stage
[Pipeline] cleanWs
12:41:15  [WS-CLEANUP] Deleting project workspace...
12:41:15  [WS-CLEANUP] Deferred wipeout is disabled by the job configuration...
12:42:07  [WS-CLEANUP] done
[Pipeline] }
[Pipeline] // timeout
[Pipeline] }
[Pipeline] // node
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] }
[Pipeline] // timeout
[Pipeline] End of Pipeline
java.lang.Exception: Error occurred during operation, please refer to logs for more information.
	at org.jfrog.build.extractor.producerConsumer.ProducerConsumerExecutor.start(ProducerConsumerExecutor.java:85)
	at org.jfrog.build.extractor.clientConfiguration.util.spec.SpecsHelper.uploadArtifactsBySpec(SpecsHelper.java:115)
	at org.jfrog.hudson.generic.GenericArtifactsDeployer$FilesDeployerCallable.invoke(GenericArtifactsDeployer.java:178)
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to cent6-x64-2
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1743)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
		at hudson.remoting.Channel.call(Channel.java:957)
		at hudson.FilePath.act(FilePath.java:1070)
		at hudson.FilePath.act(FilePath.java:1059)
		at org.jfrog.hudson.pipeline.common.executors.GenericUploadExecutor.execute(GenericUploadExecutor.java:51)
		at org.jfrog.hudson.pipeline.scripted.steps.UploadStep$Execution.run(UploadStep.java:72)
		at org.jfrog.hudson.pipeline.scripted.steps.UploadStep$Execution.run(UploadStep.java:52)
		at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:47)
		at hudson.security.ACL.impersonate(ACL.java:290)
		at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:44)
		at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
		at java.util.concurrent.FutureTask.run(Unknown Source)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
		at java.lang.Thread.run(Unknown Source)
Caused: java.lang.RuntimeException: Failed uploading artifacts by spec
	at org.jfrog.hudson.generic.GenericArtifactsDeployer$FilesDeployerCallable.invoke(GenericArtifactsDeployer.java:182)
	at org.jfrog.hudson.generic.GenericArtifactsDeployer$FilesDeployerCallable.invoke(GenericArtifactsDeployer.java:116)
	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3050)
	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:369)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/buildenv/jenkins/common/build.groovy
+++ b/buildenv/jenkins/common/build.groovy
@@ -354,8 +354,19 @@ def upload_artifactory(uploadSpec) {
     // Add BUILD_IDENTIFIER to the buildInfo. The UploadSpec adds it to the Artifact info
     buildInfo.env.filter.addInclude("BUILD_IDENTIFIER")
     buildInfo.env.capture = true
-    server.upload spec: uploadSpec, buildInfo: buildInfo
-    server.publishBuildInfo buildInfo
+
+    //Retry uploading to Artifactory if errors occur
+    def ret = false
+    retry (3) {
+        if (ret) {
+            sleep time: 300, unit: 'SECONDS'
+        } else {
+            ret = true
+        }
+        server.upload spec: uploadSpec, buildInfo: buildInfo
+        server.publishBuildInfo buildInfo
+    }
+
     // Write URL to env so that we can pull it from the upstream pipeline job
     env.ARTIFACTORY_URL = server.getUrl()
     env.ARTIFACTORY_CREDS = server.getCredentialsId()

